var/spack/repos/builtin/packages/xtensor-python/package.py

```diff
@@ -33,9 +33,4 @@ class XtensorPython(CMakePackage):
     extends("python")
 
     def cmake_args(self):
-        spec = self.spec
-
-        python_exe = spec["python"].command.path
-
-        args = ["-DPYTHON_EXECUTABLE={0}".format(python_exe)]
-        return args
+        return [f"-DPYTHON_EXECUTABLE={python.path}"]
```

Comment:
What do you mean with `-m venv`?

Reply:
Can supply a `python` executable such that when it is invoked, it automatically adds `-m venv <path>`? I thin that  would be an Executable object with `default_args=["-m venv <path>"]`

Reply:
What do you expect that to do? `python -m venv <path>` is for creating a venv on the filesystem and nothing else

Reply:
Right sorry: in that case this would look like (in `packages/python/package.py):

```
class Python():
...
    setup_dependent_package(..., dependent_spec):
        python = Executable(dependent_spec[python-venv].prefix.bin + "python")

# Likewise, we should be able to update this so that spec[python].command returns this
```

what I'm trying to avoid is a pitfall where users ask for spec[python] and it turns out to be flawed (in the sense that it doesn't use this venv we think ought to be used everywhere.

The specific `py-setuptools` version could omit all this, but just use `python -m venv ...` to create a virtualenv in the py-setuptools spack stage, then run the build using that specific virtual env. Actually, I think if we did that for every package it would fit into our current build process more idiomatically (admittedly that means it creates a venv for each install, but creating empty venvs shouldn't be that slow).

Reply:
That's what #40224 did, but was reverted

Reply:
> what I'm trying to avoid is a pitfall where users ask for spec[python] and it turns out to be flawed (in the sense that it doesn't use this venv we think ought to be used everywhere).

Yeah, that'd be great, but I feel like the only place we see `spec["python"].command.path` is in CMakePackage, in which it is in my testing redundant.

Reply:
> I don't think spec[python].command can easily transparently return spec[python-venv].command, since Python.command doesn't know from where it's being called, 

That makes sense: `.command` can be called on just `spec['python']`, not just `setup_dependent_package`.

What if:

* The `python` installation itself creates a venv as a post installation step, vs. creating it as a separate package prefix as is done now
  * Spack-installed python can then always return something reasonable for `.command`
* We add a new interface function that allows us to create a spack prefix for an external (and perform external setup steps)
  * Likewise, view logic would have to be willing to handle such packages
  * Then `python` can implement this to create a venv in the prefix

?

This way:

* Our API for providing access to python is consistent (users can still do things to cause errors, like manually constructing a path to the "wrong" python, but I think all the shortcuts we provide become consistent)
* Users can create views with `python` or `py-*` and they both work

Reply:
I've considered "installing" external python in a spack managed prefix, but I don't see how you do that. `spec.prefix` is used in many different places, and for external Python it would mean that runtime executables live in <spack prefix>/bin whereas libraries and headers live in <external prefix>/{lib,include}.

Being able to differentiate between python-venv and python is in my opinion only a good thing.

I think having a "double prefix" is going to hit many edge cases (views, build caches, uninstall) and I don't find it worth pursuing for this individual use case. The separate python-venv package has the big advantage that no internals need to be changed.

Reply:
> Being able to differentiate between python-venv and python is in my opinion only a good thing.

I feel I've brought cases in this PR where this is clearly problematic (i.e. dependents having to be careful how they access python, which is why so many packages are edited in this PR, and I think is a source of potential bugs).

> spec.prefix is used in many different places, and for external Python it would mean that runtime executables live in <spack prefix>/bin whereas libraries and headers live in <external prefix>/{lib,include}.

could we copy (like `gcc-runtime`) or symlink those into the spack prefix, so there is effectively one prefix?

My overall concern is that we are making things more difficult on the package writing side, and IMO it's worth a fair amount of pain on our end to avoid that.

Reply:
We already have a couple spec post-processing steps related to handling of external python after concretization, so perhaps this could also be handled by updating `spec[python].command` by examining the full dag and looking for `python-venv` (which would preserve your current approach). My point is I'm open to this staying the way it is if we can find a sensible way to avoid pitfalls when dependents ask for details about their `python` dependency.

Reply:
Notice that `gcc-runtime` will at some point look very similar to `python-venv`. Right now it is `gcc-runtime %gcc` but it will become `gcc-runtime ^gcc` when the compiler property is dropped, where `gcc` is likely an external package. The difference is just that you typically don't need to query `gcc-runtime` / `gcc` as is the case for `python-venv` / `python`.

I'll think a bit more about how to improve `python` command bits. I agree it's not perfect in terms of usability. In principle I do like that with the global the dependency dictates how to make it available, instead of the dependent having to figure out how to get it. It's just that those globals are magic constants and hard to trace.